### DISCUSSION TITLE
goVelocity指令の反映タイミング

### DISCUSSION THREAD

**Discussion by Naoki-Hiraoka** [YutaKojio/hrpsys-base#41] [Q&A] [answered]
goVelocity指令を見て着地位置に反映するタイミングは、一歩の切り替わりの直後 (次のfootstep_nodesの区間が始まった直後)、という理解で合っていますか? また、ビジョンからの指令に対する即応性を上げたくなった場合、一歩の途中でも常にgoVelocity指令を見て着地位置を修正するという方法が考えられますが、これは`modify_footsteps`のように実装すればできるものなのでしょうか、それとも、難しい問題があるのでしょうか。

https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L729-L732
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L754

**Comment by YutaKojio** [comment]
govelocityの指令を見ているタイミングは歩行周期の真ん中で，現在の遊脚の着地位置から反映するか次の足からにするかは overwritable_footstep_index_offset で与えています．
...だったはずなんですが，今確認したら，反映されるタイミングが一歩ずれるようになっていました．
たぶんそのときに https://github.com/YutaKojio/hrpsys-base/discussions/38 の問題も起こるようになった気がするので，合わせて確認します．

ちなみに，タイミングが毎歩一回なのは，```lcg.get_lcg_count() <= get_overwrite_check_timing()```の部分を消すだけでたぶんいつでも反映されるようになるんじゃないかと思います．
ついでに，```updated_vel_footsteps``` があるのは着地時刻延長があっても複数回更新しないようにあります．

**Reply by Naoki-Hiraoka** [reply]
> タイミングが毎歩一回なのは，lcg.get_lcg_count() <= get_overwrite_check_timing()の部分を消すだけでたぶんいつでも反映されるようになるんじゃないかと思います

なるほと、ありがとうございます。

> govelocityの指令を見ているタイミングは歩行周期の真ん中で，

`get_overwrite_check_timing()`のタイミングが、

masterだと0.5倍されていて歩行周期の真ん中に見えるのですが、
https://github.com/YutaKojio/hrpsys-base/blob/a54bca16836e43faf820dd0231ac05fd3a1d5f57/rtc/AutoBalancer/GaitGenerator.h#L1225
auto-stablizerだと1.0倍されているので歩行周期の最初に見えました
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.h#L2070

**Comment by YutaKojio** [comment]
> auto-stablizerだと1.0倍されているので歩行周期の最初に見えました

なるほど，一歩ずれてる気がしたのはそのせいか．
なんでそうしたのかすぐ思い出せないので明日それもあわせて考えます．

**Comment by YutaKojio** [comment, marked as answer]
> ...だったはずなんですが，今確認したら，反映されるタイミングが一歩ずれるようになっていました．

は，やっぱり

> auto-stablizerだと1.0倍されているので歩行周期の最初に見えました

のせいでそう見えていただけでした．
わざわざ真ん中で反映する必要がないので，バランスに影響の少ない最初のタイミングに変更していました．

また，
overwritable_footstep_index_offset=0にしたうえで，
https://github.com/YutaKojio/hrpsys-base/blob/a6f1a9f1217d43f91fe477b013d79058b85ff5d3/rtc/AutoBalancer/GaitGenerator.cpp#L754
を
```
} else {
```
に変えることで任意のタイミングで複数回着地位置を変更できることを確認しました．
ただし，着地位置が大きく変わるような変更を一歩以内で複数回指令すると転倒しました．
